### PR TITLE
feat: adopt zones in scheduler spread by constraints

### PR DESCRIPTION
**Pull request by zishen**
**What type of PR is this?**

/kind feature
/kind cleanup

**What this PR does / why we need it**:

Adopt `zones` in scheduler spread by constraints.

**Which issue(s) this PR fixes**:
part of https://github.com/karmada-io/karmada/issues/3952

**Special notes for your reviewer**:

@RainbowMango @whitewindmills @huone1 

**Does this PR introduce a user-facing change?**:

### CODE REVIEW

**Comment by zishen** [comment]
For zone is not supported yet，I do not know how to test the zone selection schedule.

see the code [here](https://github.com/karmada-io/karmada/blob/master/pkg/scheduler/core/spreadconstraint/select_clusters.go#L30-L44)

```go
func selectBestClustersBySpreadConstraints(spreadConstraints []policyv1alpha1.SpreadConstraint,
	groupClustersInfo *GroupClustersInfo, needReplicas int32) ([]*clusterv1alpha1.Cluster, error) {
	spreadConstraintMap := make(map[policyv1alpha1.SpreadFieldValue]policyv1alpha1.SpreadConstraint)
	for i := range spreadConstraints {
		spreadConstraintMap[spreadConstraints[i].SpreadByField] = spreadConstraints[i]
	}

	if _, exist := spreadConstraintMap[policyv1alpha1.SpreadByFieldRegion]; exist {
		return selectBestClustersByRegion(spreadConstraintMap, groupClustersInfo)
	} else if _, exist := spreadConstraintMap[policyv1alpha1.SpreadByFieldCluster]; exist {
		return selectBestClustersByCluster(spreadConstraintMap[policyv1alpha1.SpreadByFieldCluster], groupClustersInfo, needReplicas)
	} else {
		return nil, fmt.Errorf("just support cluster and region spread constraint")
	}
}
```

File: pkg/scheduler/core/spreadconstraint/group_clusters.go [hunk 1]
```diff
@@ -136,26 +136,29 @@ func (info *GroupClustersInfo) generateZoneInfo(spreadConstraints []policyv1alph
...
-
-		zoneInfo.Clusters = append(zoneInfo.Clusters, clusterInfo)
-		zoneInfo.AvailableReplicas += clusterInfo.AvailableReplicas
-		info.Zones[zone] = zoneInfo
 	}
 
 	for zone, zoneInfo := range info.Zones {
-		zoneInfo.Score = zoneInfo.Clusters[0].Score
+		for _, cluster := range zoneInfo.Clusters {
+			zoneInfo.Score += cluster.Score
```
Comment (RainbowMango) [inline pkg/scheduler/core/spreadconstraint/group_clusters.go:160]: I doubt the score should be summed here. 
ZoneInfo.Score indicates the highest score in all clusters of the zone. 

I think it should still be `zoneInfo.Score = zoneInfo.Clusters[0].Score`, I assume the zoneInfo.Clusters already sorted by score.

Comment (zishen) [reply pkg/scheduler/core/spreadconstraint/group_clusters.go:160]: ok

**Review by whitewindmills — COMMENTED** [review]
The zone spread constraint is still dead code for now. So I think this PR is unnecessary.

**Comment by RainbowMango** [comment]
I get it, This PR just adopted the new `zones`, how about moving this forward and you can refactor it in #3662?

**Comment by whitewindmills** [comment]
> how about moving this forward and you can refactor it in https://github.com/karmada-io/karmada/pull/3662?

these zone spread constraint codes have been deleted in https://github.com/karmada-io/karmada/pull/3662. :joy:

**Comment by RainbowMango** [comment]
I agree that dead code should be removed. 

This PR just removed the reference to deprecated code, that's the benefit we can get before #3662.
My opinion is moving this PR forward. What do you think @whitewindmills ?

**Comment by whitewindmills** [comment]
okay, just move it forward.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**